### ISSUE TITLE
bash 툴의 성공 경로 drain이 명령 자신의 출력 꼬리를 조용히 버린다 (#232의 idle 규칙)

### ISSUE THREAD

**Issue by handochan** [handochan/aelix-ai#260]
## 요약

`_LocalBashOperations.exec`의 성공 경로 drain(#232, `_drain_after_the_exit`)은 **파이프에 아직 읽을 것이 남아 있는지 묻지 않는다**. 종료 조건은 `eof` / `not reader.is_alive()` / wall-clock(idle `EXIT_DRAIN_SECONDS = 0.1`, cap)뿐이다. reader 스레드가 root의 exit을 가로질러 0.1 s 동안 스케줄되지 못하면 **커널 파이프 버퍼에 남아 있던 명령 자신의 바이트가 그대로 버려진다**. `exit_code=0`, `timed_out=False`, 잘렸다는 표시 없음.

ADR-0238 #232 amendment가 받아들인 손실은 "**백그라운드 헬퍼가 exit 뒤에 쓴 출력**"이다. 여기서 사라지는 것은 **명령 자신이 exit 전에 이미 쓴 바이트**이고, 그건 받아들인 범위 밖이다.

## 실측

CI (ubuntu-latest, py3.11, run 34238827475, `fix/249-beta2` — 이 브랜치는 `tools/`를 건드리지 않았다):

```
tests/tools/test_bash_tool_containment.py::test_every_byte_is_delivered_under_a_loaded_loop
assert 2070528 == 2097152
```

- 손실 26,624 B = 1024 B 줄로 정확히 **26줄**. 전달분 2,070,528 B = 2048줄 중 앞의 2022줄. 잘린 쪽은 **꼬리**다.
- 38,912 + 26,624 = 65,536 = 리눅스 파이프 용량 1개분. `write()`가 가득 찬 파이프에서 블록하므로 exit 순간 파이프에 남아 있을 수 있는 최대치가 정확히 그 1버퍼다. 즉 손실은 **마지막 파이프 버퍼 하나가 중간에서 잘린 것**이다.

로컬 (darwin arm64 18코어, HEAD 0985fcf):

- **자연 부하로는 재현되지 않는다.** pytest 30회(idle) / 30회(스피너 36개, load avg 33), in-process 1000 + 1000 + 500회, linux/arm64 2-vCPU 컨테이너(`--cpuset-cpus=0,1`) 500 + 300 + 300회(스피너 4개·16개 포함), 1-vCPU 200회 — 합계 약 3,300회 **0 실패**. 이 3,300회 전부에서 drain은 no-op였다: `eof`가 이미 set이었거나, 100 ms grace 중 99.8-99.9 ms를 남기고 재무장했다. **전달을 지킨 것은 drain의 타이밍이 아니라 reader 스레드가 EOF에 도달한 것이다.**
- 필요조건을 강제하면 재현된다 — **reader 스레드가 root exit을 가로질러 0.1 s 이상 GIL/CPU를 못 받는 상태** (`sys.setswitchinterval(0.5)` + 루프 스레드에서 200 ms 동기 burn). darwin **10회 중 3회 손실**, 41,984 / 41,984 / **26,624** B. 마지막은 CI 실패와 바이트 수가 같다.
- 손실 지점 계측: `eof` False, reader **ALIVE**, `FIONREAD(proc.stdout) = 26624` — **바이트는 아직 OS 파이프 안에 있었다**. `posted == delivered = 2,055,168`, 즉 loop 콜백 백로그는 **없었다**. drain의 첫 평가가 exit 스탬프로부터 400 ms 뒤에 일어났고 `last_chunk_at`은 exit보다 400 ms 앞이라 `remaining = -0.300 s` — **drain은 0 ms 기다리고 첫 눈에 break했다**.
- **루프 스레드 지연만으로는 손실이 나지 않는다.** exit→drain 진입 지연을 213-295 ms로 만들어도 darwin 0/20, linux-2cpu 0/20이고 종료 이유는 매번 `eof`다. reader가 독립적으로 EOF에 도달하기 때문이다. 프로세스 전체를 멈추는 SIGSTOP(0/40)과 187-201 ms짜리 gen-2 GC 정지(0/40)도 음성 — **loop과 reader 사이의 skew만이 바이트를 잃는다.**

CI 러너에서 reader를 굶긴 것이 정확히 무엇인지는 **미측정**이다(러너의 GIL 핸드오프인지 VM steal인지 구분하지 못했다). 위 재현은 필요조건을 인위적으로 만든 것이지 러너의 스케줄링을 재현한 것이 아니다.

## 사이트

- `packages/aelix-coding-agent/src/aelix_coding_agent/tools/bash.py:487-490`
  ```python
  while not eof.is_set() and reader.is_alive():
      armed_at = max(state.last_chunk_at, armed_from)
      remaining = min(cap, armed_at + EXIT_DRAIN_SECONDS) - time.monotonic()
      if remaining <= 0:
          break
  ```
  grace는 **exit 시점**에 무장되는데 평가는 **임의로 늦은 시점**에 일어난다. 그래서 drain이 처음 보기도 전에 예산이 이미 소진돼 있을 수 있다. 그리고 이 루프는 **파이프가 실제로 비었는지 한 번도 묻지 않는다** — non-blocking read도, `select`도, `FIONREAD`도 없다. 그러므로 "reader가 아직 안 돌았다"와 "파이프가 비었다"가 같은 관측이다.
- `packages/aelix-ai/src/aelix_ai/utils/_process_tree.py:1108` — idle 시계를 stamp하는 곳. **reader 스레드**에서 `on_chunk` 직후다. 시계 자체는 reader가 감는다. 결함은 **예산이 wall-clock**이라는 것이다.
- `packages/aelix-coding-agent/src/aelix_coding_agent/tools/bash.py:410` — `_root_exited_at`은 `proc.wait` worker 스레드에서 찍힌다. 그 스탬프와 drain 코루틴의 첫 평가 사이의 지연이 같은 0.1 s 예산에서 빠져나간다. (bash.py:405-410 주석이 이 slack을 이미 알고 있고 "13-15 ms on darwin"으로 적어 두었다 — 400 ms는 그 가정 밖이다.)
- 버려지는 두 지점, 사이에 `await`이 없다: `bash.py:631` `reader.detach()` → `_process_tree.py:1090` `if self._detached: continue` (reader 스레드에서 폐기, 시계도 안 찍음), 그리고 `bash.py:632` `delivering = False` → `bash.py:322` `if delivering:`. 위 재현에서 두 경로가 라운드에 따라 각각 관측됐다.
- 참고 — 같은 wall-clock 종료는 `run_contained`의 정상 종료 leg(`_process_tree.py:1266-1280`, #221)에도 있다. 그리고 그 함수의 docstring(`_process_tree.py:1235-1240`)이 이미 이렇게 적어 두었다: "(a) every reader has seen EOF, or is no longer alive — the ordinary end, and **the only one that proves nothing was lost**". #232는 그 문장이 "증명하지 못한다"고 말한 종료 조건을 **성공 경로**에 채택했다. 그 전(`c80e5a6^` bash.py:531)은 `await eof.wait()`였고, reader가 EOF 콜백을 마지막 청크 콜백 **뒤에** 같은 loop FIFO로 넣으므로 전달이 큐의 성질이었다.

## 사용자 비용

`create_bash_tool`의 `execute`는 `raw = b"".join(chunks)` 뒤 `truncate_tail`을 쓴다 (`bash.py:887-888`) — 그리고 `truncate_tail`은 **꼬리를 남긴다** (`tools/_truncate.py:48`, "Keep the LAST `max_lines` lines"). 잃는 바이트가 꼬리이므로 **모델에게 실제로 보여주는 바로 그 부분**이 사라진다: 실패한 assertion, 에러 요약, 마지막 상태 줄. `exit_code`는 `proc.wait`에서 오므로 멀쩡하다. 증상은 **"빌드는 실패했는데 에러가 안 보인다"**다. `info.truncated`일 때 저장하는 full-output 파일도 같은 `raw`를 쓰므로 거기에도 없다.

노출 상한은 exit 순간 파이프에 있을 수 있는 양, 즉 파이프 버퍼 1개분(리눅스 기본 64 KiB) + 진행 중인 `read1` 한 청크(`_READ_CHUNK_BYTES = 65536`)다.

출력 전체가 파이프 한 버퍼에 들어가고 즉시 exit하는 짧은 명령이면 원리상 아무것도 전달되지 않을 수 있다 — **미측정**. 위 재현이 보인 것은 2 MiB 명령의 꼬리 손실뿐이다.

빈도는 낮다. 위 3,300회가 그 증거다. 대신 **조용하다**는 것이 비용이다.

## 할 것

1. **grace를 drain이 실제로 본 시점에서도 무장한다.** `armed_at = max(state.last_chunk_at, armed_from, <drain 진입 시각>)`. 이러면 `remaining <= 0`으로 0 ms만에 break하는 퇴화가 사라지고, 기다리는 동안 loop이 selector에 park하므로 reader가 GIL을 받을 시간이 실제로 생긴다. 대가는 loop이 늦게 재개된 경우 최대 +0.1 s이고 cap이 여전히 전부를 묶는다. **위험**: `_drain_after_the_exit`은 kill leg 3개와 공유되고(bash.py:451-473) 그쪽 elapsed pin이 같이 움직인다 — kill leg는 `_end_the_tree`의 blocking `proc.wait(reap)`가 loop을 잡은 뒤에 진입하므로 그 지연이 이제 grace에 청구된다. darwin과 win32 양쪽 재측정 필요.
2. **본질: 종료 조건을 파이프의 성질로 만든다.** `eof`가 아닌 이유로 break했을 때, `detach` 전에 남은 읽을 것이 있는지 한 번 확인하고 비운다. POSIX는 `select([fd], [], [], 0)`, win32는 `PeekNamedPipe`(win32 파이프에는 `select`가 안 된다). loop 스레드에서 같은 fd를 두 번째로 읽지 않도록, reader 쪽이 "비어 있는 채로 park했다"를 stamp하게 하는 형태가 더 안전하다 — 다만 **헬퍼가 물고 있는 파이프에 park한 reader는 exit 전부터 park해 있으므로** 그대로 쓰면 #232가 산 0.14 s가 2.0 s cap으로 되돌아간다. 규칙을 그 케이스와 함께 설계할 것.
3. **EOF로 끝나지 않은 drain을 결과에 말하게 할지 결정한다.** ADR-0238이 이미 열어 둔 질문이다("What it did NOT do: say anything in the result when the drain ends on the cap rather than on EOF — Pi says nothing either, and whether Aelix should is an open product question"). 손실이 조용한 이상 최소한 표시는 있어야 한다.
4. **테스트를 약화하지 않는다.** `test_every_byte_is_delivered_under_a_loaded_loop`는 옳은 것을 재고 있다(같은 명령의 `subprocess.run` 결과와 바이트 동등). 틀린 것은 그 케이스 docstring의 전제다 (`tests/tools/test_bash_tool_containment.py:1247-1250`): "CHATTY backgrounds nothing, so the pipe EOFs at the root's own exit and the drain ends on `eof` rather than on the idle timer or the cap". 이건 reader가 grace 안에 스케줄될 때만 성립한다 — 그 문장을 고친다.
5. **ADR-0238 #232 amendment의 "What it costs, in three parts"에 이 손실을 추가한다.** 지금은 헬퍼 출력만 잘린다고 적혀 있고, 그건 불완전하다.

## 관계

- #232 / ADR-0238(#232 amendment) — 이 drain을 성공 경로에 들여온 변경. `c80e5a6`.
- #221 — `run_contained`의 같은 idle 규칙(`_process_tree.py:1266-1280`). 같은 결함 클래스가 그쪽에도 있는지는 **미측정**.
- 같은 파일의 windows 플레이크(run 34171328513, `test_a_holder_that_never_falls_idle_hits_the_drain_cap[deadline-1s]`, `assert None == 0`)는 **이 결함과 무관하고 이미 고쳐졌다** — `58f3bbb`가 win32 arm을 2.0 s로 올렸고 그 이유가 `_DEADLINE_ARM_SECONDS` 주석에 그 run 번호로 적혀 있다(`tests/tools/test_bash_tool_containment.py:872-884`). 현재 id는 `deadline-arm`이다.
- Pi #5303 / `utils/child-process.ts::waitForChildProcess` — 원래의 idle 규칙. Pi도 파이프를 묻지 않는지는 **미확인**.

**Comment by handochan** [comment]
**`b13c546c`로 닫는다(#261과 함께).** 2026-09-25 main에 머지.

## 무엇을 바꿨나

bash 도구의 drain은 이제 시계가 아니라 파이프에 대한 **증명**으로 끝난다. `run_contained`의 drain도 같다.

- **P1 `drained()`** (POSIX): reader가 들고 있는 것이 없고 파이프도 비었다.
  - reader는 non-blocking fd를 poll한다. 읽기 전에 phase를 홀수로, 넘긴 뒤에 짝수로 바꾼다.
  - 묻는 스레드는 phase가 짝수이고 FIONREAD를 읽는 동안 바뀌지 않았는지를 본다.
- **P2 `caught_up()`**: 증명 시각보다 **엄격히 뒤**에 한 첫 look이 본 것을 모두 넘겼다.
  - 이 시각은 exit·kill 스탬프를 찍는 자리에서 바로 요청한다.
  - 두 번 찍히면 **늦은 쪽**이 이긴다.
- **증명이 없을 때**: grace씩 루프를 park한다. 한도는 hard cap(exit+2.0 s, kill+1.0 s)이다.
  - look이 grace보다 더 늦고 cap 앞 grace 안이나 cap 뒤에 오면, 끝을 그 look 뒤 grace 하나로 미룬다(드레인당 한 번).
  - 증명 없이 끝나면 `output_unconfirmed`를 세우고, 모델이 받는 결과에 알림을 붙인다.
- **win32**: reader가 `read1` 직전에 **자기** 핸들을 `PeekNamedPipe`로 본다. 루프는 파이프를 절대 묻지 않는다.
- **문서**: ADR-0238에 #260 amendment를 달았고 CHANGELOG에 항목을 추가했다. `test_every_byte_is_delivered_under_a_loaded_loop`의 docstring 전제도 고쳤다(단언은 그대로다).

## 검증

- **이슈의 재현법**(`setswitchinterval(0.5)`과 루프 burn): base는 30회 중 5~13회 손실, 수정은 0회. `run_contained`도 0/10.
- **새 테스트 64개**: base에서는 전부 RED다. 결정적 seam을 쓰고, 0.3 s 정지 주입과 15.625 ms 양자화 시계에서도 통과한다.
- **라이브**(OpenRouter haiku):
  - 2 MiB 출력의 끝 마커가 모델까지 온다.
  - `(sleep 3; echo late) & echo now`는 0.11 s에 돌아온다.
  - kill 레그는 1.011 s다.
- **CI** [36017790447](https://github.com/handochan/aelix-ai/actions/runs/36017790447): 여섯 레그 모두 green, windows 두 레그 포함.

### Windows CI가 두 번 잡은 것

1. **동순위.** CPython ≤3.12 Windows의 `monotonic()`은 15.6 ms 눈금이다. 그래서 `>=` 비교에서 exit 직전의 look이 "exit 뒤"로 인정됐다. P2 증명이 거짓으로 성립해 명령 자신의 바이트를 조용히 잃을 수 있는 **제품 결함**이었다. 엄격한 `>`로 고쳤다.
2. **새 테스트 전제의 거친 시계 하한.** 이것은 테스트 쪽 문제였다.

### 교차 리뷰

- 독립 컨텍스트 Claude 리뷰 2회: must-fix 없음.
  - should-fix: 타이머에 기대는 테스트, 늦게 도는 drain의 거짓 알림.
  - nit N1: win32 kill 순서에서 무표시 손실. 이제 표시된다.
- Codex(머지 직전): POSIX 무표시 손실·pin 규칙·win32 모두 결함 없음. 발견 둘은 후속으로 옮겼다(**#340**, **#341**).

## 알고 받아들인 것

- **win32 잔여.** 마지막 빈 look 뒤에 도착한 바이트는 reader가 grace 하나 내내 굶으면 조용히 잃는다. → **#339**
- **kill 레그의 대가.** reader가 kill 시각을 가로질러 kill cap 내내 멈추면 완전한 출력에도 알림이 붙는다.
- **상한 변경.** 늦은 look이 있으면 상한은 2.0 s가 아니라 "그 look + 0.1 s"다.

## 후속

- **#338** REPL/RPC와 `run_contained` 호출자가 `output_unconfirmed`를 드러내지 않는다.
- **#339** win32 잔여.
- **#340** windows py3.11에서 park가 돈다.
- **#341** hard-cap 테스트 상한이 늦은-look 연장과 모순된다.
- **#325** 같은 drain·같은 파일, 다음 작업.
- **#332** `_drain` 시계 seam.
